### PR TITLE
Add a new tutorial: Building a Flask API server with PyTorch

### PR DESCRIPTION
**Pull request by avinassh**
I was writing this for my blog, however, I noticed that PyTorch Tutorials does not have one on building an API. So I am submitting here. My apologies, since I did not discuss this here first. I think PyTorch Tutorial could really use this 😄

### CODE REVIEW

**Comment by soumith** [comment]
@jeffreyksmithjr for review. Questions to ask:

- do you want to include this in tutorials?
- if you do, any changes you would request on the current draft of the tutorial?

**Comment by avinassh** [comment]
Hey @jeffreyksmithjr, any update on this?

**Comment by jeffreyksmithjr** [comment]
Thanks very much for thinking to add this @avinassh . I didn't see the notification before. I will review this shortly.

File: beginner_source/flask_rest_api_tutorial.py [hunk 1]
```diff
@@ -0,0 +1,336 @@
...
+#
+# - The endpoint ``/predict`` assumes that always there will be a image file
+#   in the request. This may not hold true for all requests. Our user may
+#   send image with a different parameter or send no images at all.
+#
+# - User may send non-image type files too. Since we are not handling errors,
+#   this will break our server. Loading the image within an exception block
+#   is a good idea.
+#
+# - There is a possibility that the image sent by user may not get recoginzed
```
Comment (jeffreyksmithjr) [inline beginner_source/flask_rest_api_tutorial.py]: The wording comes off a bit vague. Maybe, "Even though the model can recognize a large number of classes of images, it may not be able to recognize all images."

File: beginner_source/flask_rest_api_tutorial.py [hunk 2]
```diff
@@ -0,0 +1,336 @@
...
+#   by our model. Above code should be updated to handle this case.
```
Comment (jeffreyksmithjr) [inline beginner_source/flask_rest_api_tutorial.py]: Let's be more explicit and make a full sentence. "Enhance the implementation to handle cases when the model does not recognize anything in the image."

File: beginner_source/flask_rest_api_tutorial.py [hunk 3]
```diff
@@ -0,0 +1,336 @@
...
+#
+# - We run the Flask server in a development mode, which not suitable for
```
Comment (jeffreyksmithjr) [inline beginner_source/flask_rest_api_tutorial.py]: "server in **the** development mode, which **is** not suitable..."

File: beginner_source/flask_rest_api_tutorial.py [hunk 4]
```diff
@@ -0,0 +1,336 @@
...
+#   deploying in production. You may check `this tutorial <http://flask.pocoo.org/docs/1.0/tutorial/deploy/>`_
```
Comment (jeffreyksmithjr) [inline beginner_source/flask_rest_api_tutorial.py]: "You **can** check **out** `this tutorial`..."

File: beginner_source/flask_rest_api_tutorial.py [hunk 5]
```diff
@@ -0,0 +1,336 @@
...
+#   is a good idea.
+#
+# - There is a possibility that the image sent by user may not get recoginzed
+#   by our model. Above code should be updated to handle this case.
+#
+# - We run the Flask server in a development mode, which not suitable for
+#   deploying in production. You may check `this tutorial <http://flask.pocoo.org/docs/1.0/tutorial/deploy/>`_ 
+#   for deploying a Flask server in production.
+#
+# - You can also add a welcome page with a form which takes the image and
```
Comment (jeffreyksmithjr) [inline beginner_source/flask_rest_api_tutorial.py]: I'd phrase this slightly differently, "...add a UI by creating a page with a form..."

File: beginner_source/flask_rest_api_tutorial.py [hunk 6]
```diff
@@ -0,0 +1,336 @@
...
+#   displays the prediction. Check the `demo <https://pytorch-imagenet.herokuapp.com/>`_
```
Comment (jeffreyksmithjr) [inline beginner_source/flask_rest_api_tutorial.py]: "Check **out** the..."

File: beginner_source/flask_rest_api_tutorial.py [hunk 7]
```diff
@@ -0,0 +1,336 @@
...
+# - There is a possibility that the image sent by user may not get recoginzed
+#   by our model. Above code should be updated to handle this case.
+#
+# - We run the Flask server in a development mode, which not suitable for
+#   deploying in production. You may check `this tutorial <http://flask.pocoo.org/docs/1.0/tutorial/deploy/>`_ 
+#   for deploying a Flask server in production.
+#
+# - You can also add a welcome page with a form which takes the image and
+#   displays the prediction. Check the `demo <https://pytorch-imagenet.herokuapp.com/>`_ 
+#   of the same or its `source code <https://github.com/avinassh/pytorch-flask-api-heroku>`_.
```
Comment (jeffreyksmithjr) [inline beginner_source/flask_rest_api_tutorial.py]: I'd rephrase to, "of **a similar project and its**..."

**Comment by jeffreyksmithjr** [comment]
Thanks very much @avinassh  for this well-conceived review. I have a bunch of small things around clarity and language. Overall, this is a valuable contribution. Once we've got it wrapped up, I want to point to this from other locations as well, to help PT users find this resource.

**Comment by jspisak** [comment]
@avinassh - any update on this one? We'd like to merge this by next week for the Summer Hack..

cc @brianjo @jlin27

**Comment by avinassh** [comment]
somehow I missed the first notification. Really sorry. I will update the PR now.

File: beginner_source/README.txt [hunk 8]
```diff
@@ -20,3 +20,7 @@ Beginner Tutorials
 5. nlp/* and deep_learning_nlp_tutorial.rst
 	Deep Learning for NLP with Pytorch
 	https://pytorch.org/tutorials/beginner/deep_learning_nlp_tutorial.html
+
+6. flask_rest_api_tutorial.py
```
[the remark this answers is not in the document]
Comment (avinassh) [reply beginner_source/README.txt]: Sure. Moved to the intermediate section.

**Comment by avinassh** [comment]
@jeffreyksmithjr Thank you so much for the PR review. Since I am a non-native English speaker, I ended up making a lot of grammatical mistakes. 

I have fixed the PR as per the review suggestions. 

Only one thing left is about transforms mean/std deviation magic numbers - https://github.com/pytorch/tutorials/pull/537#discussion_r299577085

I have provided a link to torchvision's [model doc](https://pytorch.org/docs/stable/torchvision/models.html), but it doesn't explain either how/why about those numbers

Could you suggest me, what to add?

**Comment by avinassh** [comment]
@jspisak I have updated the PR :D

**Comment by jspisak** [comment]
@avinassh - thanks!!  @jeffreyksmithjr - take a look and merge?

File: intermediate_source/flask_rest_api_tutorial.py [hunk 9]
```diff
@@ -204,12 +207,12 @@ def get_prediction(image_bytes):
 # readable name.
 #
 # .. Note ::
-#    Did you notice why ``model`` variable is not part of ``get_prediction``
-#    method? Or why it is a global variable? Loading a model can be an
-#    expensive operation, memory, CPU wise. If we loaded the model in the
-#    ``get_prediction`` method, then it will get unnecessarily loaded every
+#    Did you notice that why ``model`` variable is not part of ``get_prediction``
```
Comment (jeffreyksmithjr) [inline intermediate_source/flask_rest_api_tutorial.py]: `why` still is incorrect. It should just be `Did you notice that model..`

**Comment by jspisak** [comment]
Merging per Jeff's go ahead..

**Comment by jeffreyksmithjr** [comment]
I think this a great tutorial, and I really appreciate you writing it @avinassh . I appreciate your patience with my various nits around writing style. Thank you for taking my feedback and getting this in.

File: intermediate_source/flask_rest_api_tutorial.py (hunk 9 above)
Comment (avinassh) [reply intermediate_source/flask_rest_api_tutorial.py]: oops, somehow I missed it. Sent another PR #583 to fix it.

**Comment by avinassh** [comment]
Thank you so much @jeffreyksmithjr! It was really fun writing it. Thanks for your feedback, it was very helpful and I learnt quite a few things about writing a good tutorial :D

**Comment by avinassh** [comment]
Also, may I know when this will go live on https://pytorch.org/tutorials

**Comment by soumith** [comment]
it should be live now, as soon as the CI finishes the job and pushes to the site.
